arelle/plugin/validate/ROS/rules/ros.py

```diff
@@ -354,9 +354,9 @@ def rule_ros19(
     message: str | None = None
     modelObjects = set()
     monetaryFacts = val.modelXbrl.factsByDatatype(False, qnXbrliMonetaryItemType)
-    monetaryUnits = set(fact.unit.value for fact in monetaryFacts)
+    monetaryUnits = set(fact.unit.value if fact.unit and fact.unit.value else None for fact in monetaryFacts)
     pricipalCurrencyFacts = val.modelXbrl.factsByLocalName.get(PRINCIPAL_CURRENCY, set())
-    principalCurrencyValues = set(fact.text for fact in pricipalCurrencyFacts)
+    principalCurrencyValues = set(fact.text if fact.text else None for fact in pricipalCurrencyFacts)
```

Comment:
Why this change?  If xsi:nil or empty string, it seems like those _should_ be added to the set to force a mismatch.